### PR TITLE
Use log probabilities in beam search.

### PR DESCRIPTION
**Pull request by a-sneddon**
Implemented log-space operations in beam search to avoid underflow when dealing with regular probabilities.

### CODE REVIEW

File: ctc_decoder/beam_search.py [hunk 1]
```diff
@@ -42,8 +45,11 @@ def apply_lm(parent_beam, child_beam, labels, lm):
         c1 = labels[parent_beam.labeling[-1] if parent_beam.labeling else labels.index(' ')]  # first char
         c2 = labels[child_beam.labeling[-1]]  # second char
         lm_factor = 0.01  # influence of language model
-        bigram_prob = lm.get_char_bigram(c1, c2) ** lm_factor
-        child_beam.pr_text = parent_beam.pr_text * bigram_prob  # probability of char sequence
+        bigram_prob = lm_factor * np.log(lm.get_char_bigram(c1, c2))
+        if parent_beam.is_empty():
+            child_beam.pr_text = bigram_prob  # first char in beam
```
Comment (githubharald) [inline ctc_decoder/beam_search.py:50]: no special case needed (no if else), as an empty beam is initialized with pr_text=log(1)=0, so adding parent_beam.pr_text + bigram_prob is OK

File: ctc_decoder/beam_search.py [hunk 2]
```diff
@@ -75,8 +81,8 @@ def beam_search(mat: np.ndarray, labels: str, beam_width: int = 25, lm: Optional
     last = BeamState()
     labeling = ()
     last.entries[labeling] = BeamEntry()
-    last.entries[labeling].pr_blank = 1
-    last.entries[labeling].pr_total = 1
+    last.entries[labeling].pr_blank = LOG_ZERO
```
Comment (githubharald) [inline ctc_decoder/beam_search.py:84]: should be log(1)==0 instead of LOG_ZERO which is -inf

File: ctc_decoder/beam_search.py [hunk 3]
```diff
@@ -89,23 +95,29 @@ def beam_search(mat: np.ndarray, labels: str, beam_width: int = 25, lm: Optional
         for labeling in best_labelings:
 
             # probability of paths ending with a non-blank
-            pr_non_blank = 0
+            pr_non_blank = LOG_ZERO
             # in case of non-empty beam
             if labeling:
                 # probability of paths with repeated last char at the end
-                pr_non_blank = last.entries[labeling].pr_non_blank * mat[t, labeling[-1]]
```
Comment (githubharald) [inline ctc_decoder/beam_search.py:96]: to keep the code as uncluttered as possible, I suggest moving the logic which takes care of adding two values together into a separate function, as you need it multiple times and we can get rid of the if / else in the code.
Something like:
```

def log_add(x, y):
    if x == log(0):
        return y
    if y == log(0):
        return x
    return x + y
```

File: ctc_decoder/beam_search.py [hunk 4]
```diff
@@ -89,23 +95,29 @@ def beam_search(mat: np.ndarray, labels: str, beam_width: int = 25, lm: Optional
...
+                if last.entries[labeling].pr_non_blank == LOG_ZERO:
+                    pr_non_blank = np.log(mat[t, labeling[-1]])  # cannot add to -inf
+                else:
+                    pr_non_blank = last.entries[labeling].pr_non_blank + np.log(mat[t, labeling[-1]])
 
             # probability of paths ending with a blank
-            pr_blank = last.entries[labeling].pr_total * mat[t, blank_idx]
+            if last.entries[labeling].pr_total == LOG_ZERO:
```
Comment (githubharald) [inline ctc_decoder/beam_search.py:108]: same as above

File: ctc_decoder/beam_search.py [hunk 5]
```diff
@@ -89,23 +95,29 @@ def beam_search(mat: np.ndarray, labels: str, beam_width: int = 25, lm: Optional
...
+                pr_blank = np.log(mat[t, blank_idx]) # cannot add to -inf
```
Comment (githubharald) [inline ctc_decoder/beam_search.py:109]: directly using np.log creates warning outputs for np.log(0).
Define some wrapper function which takes care of that:
```
def log(x):
    with np.errstate(divide='ignore'):
        return np.log(x)
```

**Comment by githubharald** [comment]
thanks for your contribution. Using log-probs is a good idea 👍.
There are a few minor things I suggest to change, to have as few if-else branches directly in the main decoding function, and to avoid having numpy print warning messages to the console.

**Comment by githubharald** [comment]
I anyway will push some changes to the repo, so I'll integrate your changes and will do the small modifications myself.
Thanks again for your contribution 👍 .

File: ctc_decoder/beam_search.py (hunk 1 above)
Comment (a-sneddon) [reply ctc_decoder/beam_search.py:50]: Thanks for picking this up.

File: ctc_decoder/beam_search.py (hunk 2 above)
Comment (a-sneddon) [reply ctc_decoder/beam_search.py:84]: That's true, that allows the below if statements to be cleaned up also, thanks for picking this up.

**Comment by a-sneddon** [comment]
Thanks @githubharald for accepting the pull request.  I've taken a look at the changes you've implemented since and they look good, thanks for your improvements and for maintaining the repo.

**Pull request merged**